File: apps/android/app/src/main/java/com/smartble/core/model/BleDevice.kt
package com.smartble.core.model

/**
 * BLE 设备
 */
data class BleDevice(
    val id: String,
    val name: String?,
    val rssi: Int,
    val state: ConnectionState = ConnectionState.Disconnected,
    val scanRecord: ScanRecord? = null,
) {
    val displayName: String
        get() = name ?: "未知设备"

    val rssiLevel: RssiLevel
        get() = when {
            rssi >= -50 -> RssiLevel.Excellent
            rssi >= -70 -> RssiLevel.Good
            rssi >= -90 -> RssiLevel.Fair
            else -> RssiLevel.Weak
        }
}

/**
 * 连接状态
 */
enum class ConnectionState {
    Disconnected,
    Connecting,
    Connected,
    Disconnecting,
}

/**
 * 信号强度等级
 */
enum class RssiLevel {
    Excellent,
    Good,
    Fair,
    Weak,
}

/**
 * 扫描记录
 */
data class ScanRecord(
    val serviceUuids: List<String>? = null,
    val manufacturerData: Map<Int, ByteArray>? = null,
    val serviceData: Map<String, ByteArray>? = null,
    val txPowerLevel: Int? = null,
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ScanRecord

        if (serviceUuids != other.serviceUuids) return false
        if (manufacturerData != other.manufacturerData) return false
        if (serviceData != other.serviceData) return false
        if (txPowerLevel != other.txPowerLevel) return false

        return true
    }

    override fun hashCode(): Int {
        var result = serviceUuids?.hashCode() ?: 0
        result = 31 * result + (manufacturerData?.hashCode() ?: 0)
        result = 31 * result + (serviceData?.hashCode() ?: 0)
        result = 31 * result + (txPowerLevel ?: 0)
        return result
    }
}

/**
 * 扫描结果
 */
data class ScanResult(
    val device: BleDevice,
    val timestampNanos: Long,
)

File: apps/flutter/lib/core/models/ble_device.dart
/// BLE 设备状态
enum BleDeviceState {
  /// 未连接
  disconnected,
  /// 连接中
  connecting,
  /// 已连接
  connected,
  /// 断开中
  disconnecting,
}

/// BLE 设备模型
class BleDevice {
  /// 设备 ID（MAC 地址或平台特定标识）
  final String id;

  /// 设备名称
  final String name;

  /// 信号强度 (dBm)
  final int rssi;

  /// 广播数据
  final List<int>? advertisData;

  /// 服务 UUID 列表
  final List<String> serviceUuids;

  /// 连接状态
  final BleDeviceState state;

  /// 最后发现时间
  final DateTime lastSeen;

  /// 是否连接
  bool get isConnected => state == BleDeviceState.connected;

  /// 是否正在连接
  bool get isConnecting => state == BleDeviceState.connecting;

  /// 显示名称（优先使用 name，否则显示设备 ID）
  String get displayName => name.isNotEmpty ? name : '未知设备 ($id)';

  const BleDevice({
    required this.id,
    required this.name,
    required this.rssi,
    this.advertisData,
    this.serviceUuids = const [],
    this.state = BleDeviceState.disconnected,
    required this.lastSeen,
  });

  /// 复制并修改部分属性
  BleDevice copyWith({
    String? id,
    String? name,
    int? rssi,
    List<int>? advertisData,
    List<String>? serviceUuids,
    BleDeviceState? state,
    DateTime? lastSeen,
  }) {
    return BleDevice(
      id: id ?? this.id,
      name: name ?? this.name,
      rssi: rssi ?? this.rssi,
      advertisData: advertisData ?? this.advertisData,
      serviceUuids: serviceUuids ?? this.serviceUuids,
      state: state ?? this.state,
      lastSeen: lastSeen ?? this.lastSeen,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is BleDevice && runtimeType == other.runtimeType && id == other.id;

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() {
    return 'BleDevice(id: $id, name: $name, rssi: $rssi, state: $state)';
  }
}
